vernac/library.ml

```diff
@@ -248,47 +249,77 @@ let mk_summary m = {
   libsum_info = m.library_info;
 }
 
-let mk_intern_library sum lib digest_lib proofs =
+let mk_intern_library sum lib digest_lib proofs vm =
   add_opaque_table sum.md_name (ToFetch proofs);
   let open Safe_typing in
-  mk_library sum lib (Dvo_or_vi digest_lib)
+  mk_library sum lib (Dvo_or_vi digest_lib) vm
 
 let summary_seg : seg_sum ObjFile.id = ObjFile.make_id "summary"
 let library_seg : seg_lib ObjFile.id = ObjFile.make_id "library"
 let opaques_seg : seg_proofs ObjFile.id = ObjFile.make_id "opaques"
 let vm_seg : seg_vm ObjFile.id = Vmlibrary.vm_segment
 
-let intern_from_file ?loc lib_resolver dir =
-  let f = lib_resolver dir in
-  Feedback.feedback(Feedback.FileDependency (Some f, DirPath.to_string dir));
-  let ch = raw_intern_library f ?loc in
+module Intern = struct
+  module Provenance = struct
+    type t = string * string
+    (** A pair of [kind, object], for example ["file",
+        "/usr/local/foo.vo"], used for error messages. *)
+  end
+  module Error = struct
+    type t = Pp.t
+  end
+
+  type t = DirPath.t -> (library_t, Error.t) Result.t * Provenance.t
+end
+
+let intern_from_file file =
+  let ch = raw_intern_library file in
   let lsd, digest_lsd = ObjFile.marshal_in_segment ch ~segment:summary_seg in
   let lmd, digest_lmd = ObjFile.marshal_in_segment ch ~segment:library_seg in
-  let del_opaque, _ = in_delayed f ch ~segment:opaques_seg in
-  let vmlib = Vmlibrary.load dir ~file:f ch in
+  let del_opaque, _ = in_delayed file ch ~segment:opaques_seg in
+  let vmlib = Vmlibrary.load lsd.md_name ~file ch in
   ObjFile.close_in ch;
-  System.check_caml_version ~caml:lsd.md_ocaml ~file:f;
-  register_library_filename lsd.md_name f;
-  (* [dir] is an absolute name which matches [f] which must be in loadpath *)
-  if not (DirPath.equal dir lsd.md_name) then
-    user_err ?loc
-      (str "The file " ++ str f ++ str " contains library" ++ spc () ++
-       DirPath.print lsd.md_name ++ spc () ++ str "and not library" ++
-       spc() ++ DirPath.print dir ++ str ".");
-  Feedback.feedback (Feedback.FileLoaded(DirPath.to_string dir, f));
+  System.check_caml_version ~caml:lsd.md_ocaml ~file;
+  register_library_filename lsd.md_name file;
   Library_info.warn_library_info ~transitive:true lsd.md_name lsd.md_info;
-  lsd, lmd, digest_lmd, del_opaque, vmlib
+  mk_intern_library lsd lmd digest_lmd del_opaque vmlib
 
+let intern_from_file file =
+  let provenance = ("file", file) in
+  (match CErrors.to_result ~f:intern_from_file file with
+   | Ok res -> Ok res
+   | Error iexn -> Error (CErrors.iprint iexn)), provenance
```

Comment:
Seems like the wrong place to print, we should return the exn (and error_in_intern wraps it in `exception InternError of exn * dir * ...`)

Reply:
Yup, I didn't like that, but was exhausted. Thanks for the suggestion, will revise when the dep PRs are merged.

Reply:
Tried to address the comment in the last commit.